[4/n][dagster-sling] Mark DagsterSlingTranslator.get_* methods as superseded

maximearmstrong opened the pull request:
## Summary & Motivation

As title.

maximearmstrong commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/27142?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#27142** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/27142?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/27142?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#27141** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/27141?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#27140** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/27140?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#27139** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/27139?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

cmpadden commented:
@maximearmstrong , I'm curious if it's necessary to deprecate the `get_` methods, with `get_asset_spec` being the favored approach. Is this being standardized across all integrations, dbt, dlt, etc?

I just want to be cognizant of the impact it has on our users who having working pipelines, and want them to continue to work with little to no changes.

OwenKephart commented:
I agree with @cmpadden 's points here -- I think maybe `superseded` is a better decorator here? At the very least the deprecation cycle should be longer

I think the maintenance burden of these extra methods isn't too high on our side (even if they're a bit unsightly)

OwenKephart reviewed (changes requested):
Following up now that I've looked at the downstack PR, I think there's a way to delete these methods without breaking users and without having to keep these in code forever

Also having thought about it more, I'm not sure there's a good way for us to actually mark these as deprecated even if we wanted to -- the user is overriding these functions not calling them, so marking them this way won't really alert the user that anything is "wrong"

maximearmstrong commented:
> Also having thought about it more, I'm not sure there's a good way for us to actually mark these as deprecated even if we wanted to -- the user is overriding these functions not calling them, so marking them this way won't really alert the user that anything is "wrong"

@OwenKephart I updated the decorator to mark methods as superseded. I think keeping the decorators and methods are important:

- This will ensure that `get_*` methods are marked as superseded in the docs. 
- These methods can be manually called by users.
  - Nothing prevents users to do something like `DagsterSlingTranslator().get_asset_key(stream_definition)` in their code to recreate an asset key - it's not best practice, but not forbidden.

To make sure users get the supersession warning, I updated `_resolve_back_compat_method` to emit a warning if a `get_*` method is used.

maximearmstrong commented:
> I'm curious if it's necessary to deprecate the `get_` methods, with `get_asset_spec` being the favored approach. Is this being standardized across all integrations, dbt, dlt, etc?

@cmpadden This [thread](https://github.com/dagster-io/dagster/pull/25751#discussion_r1830062563) can give you more context on why we think `get_` methods should be deprecated. 

The main problem is that users can override `get_asset_key` without using the correct and new implementation in `get_asset_spec`. This caused many problems in the past, where users would have asset key with different prefixes, or asset duplication, because of that.

The conclusion was that only `get_asset_spec` should be implemented, and then attributes could be accessed from the `AssetSpec` object.

OwenKephart commented on python_modules/libraries/dagster-sling/dagster_sling/dagster_sling_translator.py:
```diff
@@ -60,6 +61,12 @@ def _resolve_back_compat_method(
+            supersession_warning(
+                subject=method_name,
+                additional_warn_text=(
+                    "Use `DagsterSlingTranslator.get_asset_spec(...)` to access the attribute instead."
```
so this warning comes up specifically in cases where users are overriding the method on the base class, so maybe the warning should reference that? i.e. `Instead of overriding X, override DagsterSlingTranslator.get_asset_spec()`

OwenKephart reviewed (approved):
Note my comment on the warning message
